File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/PIDFController.java
package org.firstinspires.ftc.teamcode.util;

import androidx.annotation.Nullable;

/**
 * PID controller with various feedforward components.
 */
public final class PIDFController {
    public static final class PIDCoefficients {
        public double kP, kI, kD;

        public PIDCoefficients(double p, double i, double d){
            this.kP = p;
            this.kI = i;
            this.kD = d;

        }
    }

    public interface FeedforwardFun {
        double compute(double position, @Nullable Double velocity);
    }

    private final PIDCoefficients pid;
    private final double kV, kA, kStatic;
    private final FeedforwardFun kF;

    private double errorSum;
    private long lastUpdateTs;

    private boolean inputBounded;
    private double minInput, maxInput;

    private boolean outputBounded;
    private double minOutput, maxOutput;

    /**
     * Target position (that is, the controller setpoint).
     */
    public double targetPosition;

    /**
     * Target velocity.
     */
    public double targetVelocity;

    /**
     * Target acceleration.
     */
    public double targetAcceleration;

    /**
     * Error computed in the last call to {@link #update(long, double, Double)}
     */
    public double lastError;

    /**
     * Feedforward parameters {@code kV}, {@code kA}, and {@code kStatic} correspond with a basic
     * kinematic model of DC motors. The general function {@code kF} computes a custom feedforward
     * term for other plants.
     *
     * @param pid traditional PID coefficients
     * @param kV feedforward velocity gain
     * @param kA feedforward acceleration gain
     * @param kStatic additive feedforward constant
     * @param kF custom feedforward that depends on position and velocity
     */
    public PIDFController(
            PIDCoefficients pid,
            double kV,
            double kA,
            double kStatic,
            FeedforwardFun kF
    ) {
        this.pid = pid;
        this.kV = kV;
        this.kA = kA;
        this.kStatic = kStatic;
        this.kF = kF;
    }

    public PIDFController(
            PIDCoefficients pid,
            double kV,
            double kA,
            double kStatic
    ) {
        this(pid, kV, kA, kStatic, (x, v) -> 0);
    }

    public PIDFController(
            PIDCoefficients pid,
            FeedforwardFun kF
    ) {
        this(pid, 0, 0, 0, kF);
    }

    public PIDFController(
            PIDCoefficients pid
    ) {
        this(pid, 0, 0, 0);
    }

    /**
     * Sets bound on the input of the controller. When computing the error, the min and max are
     * treated as the same value. (Imagine taking the segment of the real line between min and max
     * and attaching the endpoints.)
     *
     * @param min minimum input
     * @param max maximum input
     */
    public void setInputBounds(double min, double max) {
        if (min < max) {
            inputBounded = true;
            minInput = min;
            maxInput = max;
        }
    }

    /**
     * Sets bounds on the output of the controller.
     *
     * @param min minimum output
     * @param max maximum output
     */
    public void setOutputBounds(double min, double max) {
        if (min < max) {
            outputBounded = true;
            minOutput = min;
            maxOutput = max;
        }
    }

    private double getPositionError(double measuredPosition) {
        double error = targetPosition - measuredPosition;
        if (inputBounded) {
            final double inputRange = maxInput - minInput;
            while (Math.abs(error) > inputRange / 2.0) {
                error -= Math.copySign(inputRange, error);
            }
        }
        return error;
    }

    /**
     * Run a single iteration of the controller.
     *
     * @param timestamp measurement timestamp as given by {@link System#nanoTime()}
     * @param measuredPosition measured position (feedback)
     * @param measuredVelocity measured velocity
     */
    public double update(
            long timestamp,
            double measuredPosition,
            @Nullable Double measuredVelocity
    ) {
        final double error = getPositionError(measuredPosition);

        if (lastUpdateTs == 0) {
            lastError = error;
            lastUpdateTs = timestamp;
            return 0;
        }

        final double dt = timestamp - lastUpdateTs;
        errorSum += 0.5 * (error + lastError) * dt;
        final double errorDeriv = (error - lastError) / dt;

        lastError = error;
        lastUpdateTs = timestamp;

        double velError;
        if (measuredVelocity == null) {
            velError = errorDeriv;
        } else {
            velError = targetVelocity - measuredVelocity;
        }

        double baseOutput = pid.kP * error + pid.kI * errorSum + pid.kD * velError +
                kV * targetVelocity + kA * targetAcceleration +
                kF.compute(measuredPosition, measuredVelocity);

        double output = 0;
        if (Math.abs(baseOutput) > 1e-6) {
            output = baseOutput + Math.copySign(kStatic, baseOutput);
        }

        if (outputBounded) {
            return Math.max(minOutput, Math.min(output, maxOutput));
        }

        return output;
    }

    public double update(
            long timestamp,
            double measuredPosition
    ) {
        return update(timestamp, measuredPosition, null);
    }

    public double update(
            double measuredPosition
    ) {
        return update(System.nanoTime(), measuredPosition, null);
    }

    /**
     * Reset the controller's integral sum.
     */
    public void reset() {
        errorSum = 0;
        lastError = 0;
        lastUpdateTs = 0;
    }
}

File: RoadRunner/src/main/kotlin/com/acmerobotics/roadrunner/control/PIDFController.kt
package com.acmerobotics.roadrunner.control

import com.acmerobotics.roadrunner.util.NanoClock
import com.acmerobotics.roadrunner.util.epsilonEquals
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sign

/**
 * PID controller with various feedforward components.
 */
class PIDFController
/**
 * Constructor for [PIDFController]. [kV], [kA], and [kStatic] are designed for DC motor feedforward
 * control (the most common kind of feedforward in FTC). [kF] provides a custom feedforward term for other plants.
 *
 * @param pid traditional PID coefficients
 * @param kV feedforward velocity gain
 * @param kA feedforward acceleration gain
 * @param kStatic additive feedforward constant
 * @param kF custom feedforward that depends on position and/or velocity (e.g., a gravity term for arms)
 * @param clock clock
 */
@JvmOverloads constructor(
    private val pid: PIDCoefficients,
    private val kV: Double = 0.0,
    private val kA: Double = 0.0,
    private val kStatic: Double = 0.0,
    private val kF: (Double, Double?) -> Double = { _, _ -> 0.0 },
    private val clock: NanoClock = NanoClock.system()
) {
    private var errorSum: Double = 0.0
    private var lastUpdateTimestamp: Double = Double.NaN

    private var inputBounded: Boolean = false
    private var minInput: Double = 0.0
    private var maxInput: Double = 0.0

    private var outputBounded: Boolean = false
    private var minOutput: Double = 0.0
    private var maxOutput: Double = 0.0

    /**
     * Target position (that is, the controller setpoint).
     */
    var targetPosition: Double = 0.0

    /**
     * Target velocity.
     */
    var targetVelocity: Double = 0.0

    /**
     * Target acceleration.
     */
    var targetAcceleration: Double = 0.0

    /**
     * Error computed in the last call to [update].
     */
    var lastError: Double = 0.0
        private set

    /**
     * Sets bound on the input of the controller. The min and max values are considered modularly-equivalent (that is,
     * the input wraps around).
     *
     * @param min minimum input
     * @param max maximum input
     */
    fun setInputBounds(min: Double, max: Double) {
        if (min < max) {
            inputBounded = true
            minInput = min
            maxInput = max
        }
    }

    /**
     * Sets bounds on the output of the controller.
     *
     * @param min minimum output
     * @param max maximum output
     */
    fun setOutputBounds(min: Double, max: Double) {
        if (min < max) {
            outputBounded = true
            minOutput = min
            maxOutput = max
        }
    }

    private fun getPositionError(measuredPosition: Double): Double {
        var error = targetPosition - measuredPosition
        if (inputBounded) {
            val inputRange = maxInput - minInput
            while (abs(error) > inputRange / 2.0) {
                error -= sign(error) * inputRange
            }
        }
        return error
    }

    /**
     * Run a single iteration of the controller.
     *
     * @param measuredPosition measured position (feedback)
     * @param measuredVelocity measured velocity
     */
    @JvmOverloads
    fun update(
        measuredPosition: Double,
        measuredVelocity: Double? = null
    ): Double {
        val currentTimestamp = clock.seconds()
        val error = getPositionError(measuredPosition)
        return if (lastUpdateTimestamp.isNaN()) {
            lastError = error
            lastUpdateTimestamp = currentTimestamp
            0.0
        } else {
            val dt = currentTimestamp - lastUpdateTimestamp
            errorSum += 0.5 * (error + lastError) * dt
            val errorDeriv = (error - lastError) / dt

            lastError = error
            lastUpdateTimestamp = currentTimestamp

            // note: we'd like to refactor this with Kinematics.calculateMotorFeedforward() but kF complicates the
            // determination of the sign of kStatic
            val baseOutput = pid.kP * error + pid.kI * errorSum +
                pid.kD * (measuredVelocity?.let { targetVelocity - it } ?: errorDeriv) +
                kV * targetVelocity + kA * targetAcceleration + kF(measuredPosition, measuredVelocity)
            val output = if (baseOutput epsilonEquals 0.0) 0.0 else baseOutput + sign(baseOutput) * kStatic

            if (outputBounded) {
                max(minOutput, min(output, maxOutput))
            } else {
                output
            }
        }
    }

    /**
     * Reset the controller's integral sum.
     */
    fun reset() {
        errorSum = 0.0
        lastError = 0.0
        lastUpdateTimestamp = Double.NaN
    }
}
